services/statestorage/service.go

```diff
@@ -55,12 +56,18 @@ func (s *service) CommitStateDiff(input *services.CommitStateDiffInput) (*servic
 
 	s.reporting.Info("trying to commit state diff", log.BlockHeight(commitBlockHeight), log.Int("number-of-state-diffs", len(input.ContractStateDiffs)))
 
-	if lastCommittedBlock := s.lastCommittedBlockHeader.BlockHeight(); lastCommittedBlock+1 != commitBlockHeight {
+	lastCommittedBlock := s.lastCommittedBlockHeader.BlockHeight()
+	if  lastCommittedBlock+1 != commitBlockHeight {
 		return &services.CommitStateDiffOutput{NextDesiredBlockHeight: lastCommittedBlock + 1}, nil
 	}
 
 	// if updating state records fails downstream the merkle tree entries will not bother us
-	s.merkle.Update(input.ContractStateDiffs)
+	if root, err := s.persistence.ReadMerkleRoot(commitBlockHeight-1); err != nil { // TODO use input.resultheader.preexecutuion
+		return nil, errors.Errorf("cannot find previous block merkle root. current block %d", commitBlockHeight)
+	} else {
```

Comment:
no need for else here. but it would require declaring root and err before the if.